### ISSUE TITLE
automated PUSH FAILED messages coming from Rosetta

### ISSUE THREAD

**Issue by jbphet** [phetsims/rosetta#105]
This issue was originally reported in perennial, see https://github.com/phetsims/perennial/issues/5, moved here because it's really a rosetta issue.

@samreid said:

I received this PUSH FAILED email on the aforementioned date:

> Error: Repo createContents error
> 
> Error: Repo createContents error. Error committing to file acid-base-solutions/acid-base-solutions-strings_te.json
> 
> [object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]

Assigning @jbphet since it is related to build-server.  Tagging @pixelzoom since it relates to acid-base-solutions.

**Comment by jbphet** [comment]
@pixelzoom said:

The `[object Object]...` portion of that error message is pretty useless.  A related issue would be to improve the error message.

**Comment by jbphet** [comment]
Gathering information on this.  The commit in question seems to have made it into babel, and the timestamp matches that reported in the email message, i.e. 11/25/15, 2:09 PM MST.  See https://github.com/phetsims/babel/commit/f293f542d524e155429f68a1dc40b74146cb37d8.

**Comment by jbphet** [comment]
The translation was successfully posted, see http://phet.colorado.edu/sims/html/acid-base-solutions/latest/acid-base-solutions_te.html.  The date information on the file on figaro looks about right, meaning that it was just a few minutes after 2:09 PM on 11/25/15.

**Comment by jbphet** [comment]
I spent quite a bit of time combing through rosetta's log and looking at the code.  My best forensic analysis at this point is this:
+ Rosetta tried to created the translated string file acid-base-solutions-strings_te.json in the ```babel``` repo using the ```octonode``` npm package, and received an error message, which was logged
+ Rosetta then set a 5 second timeout and tried again
+ The second attempt to create the new string file also received a failure response, which was logged, but it looks like the code in Rosetta handles the second failure as though it succeeded, which is extremely suspect
+ In the handling of the 2nd (failed) attempt to create the string file, Rosetta sends a request to the build server to build the translated simulation, which again is highly suspect since this build should fail if the submission of the translated strings actually failed, but it appears that the build actually worked, and that the creation of the string file actually worked despite the error message
+ Rosetta sent out the PUSH FAILED email because of the two apparent failures to create the string file

There are clearly some problems here.  My guess is that this was happening occasionally during the testing of rosetta, and the original author of the code decided to experiment with treating the 2nd failed creation attempt as a success to see if there really was a failure.  This worked, but never got cleaned up.  So right now we have some code that appears to work in this particular case, sends out incorrect error messages, and relies on partially ignoring an error message.  This obviously needs some work.  I'll work with @ariel-phet to determine how this should be prioritized.

**Comment by ariel-phet** [comment]
@jbphet and I discussed over skype on 12/1/2015. Since no issues have been reported by translators we do not feel this issue is a high priority, but it would be good to clean up as the interaction is suspect.

Currently, the plan is to work on this in January.

**Comment by jbphet** [comment]
I made some improvements to the contents of the email message so that it should print out the string fields rather than the useless ```[object Object]``` stuff that it was doing previously.  This should be checked the next time this error condition is encountered.

**Comment by jbphet** [comment]
Another such email was received on 12/9/2015 at 2:53 AM local (Mountain Standard) time.  Again, it looks like the push ultimately succeeded, so this email was incorrect.  The improvements to the email message (described in previous comment) appear to have worked, and here is what the email looked like:

> From: ?= <PhET Build Server <phethelp@colorado.edu>
> Subject: PUSH FAILED
>
> Error: Repo updateContents error
>
> Error: Repo updateContents error. Error committing to file scenery-phet/scenery-phet-strings_da.json
>
> [{"stringKey":"FaceWithPointsNode.pattern_0sign_1points","stringValue":"{0}{1}"}]

Notice that the ```[object Object]``` stuff has been replaced with actual string key values.

**Comment by samreid** [comment]
I also saw the "From: ?= " on my phone, but From: PhET Build Server <phethelp@colorado.edu> on my laptop.

Any idea why it is sending an error message when the commit succeeded?  Or is that for subsequent investigation?

**Comment by jbphet** [comment]
@samreid - the cause of the issue is explained here: https://github.com/phetsims/rosetta/issues/105#issuecomment-160797990.  These is definitely some room for improvement in how Rosetta handles error legs when interacting with GitHub.  This will be worked on in the future.

**Comment by jbphet** [comment]
Another one of these messages was received on 12/18/2015.  Contents were:
> Error: Repo updateContents error

> Error: Repo updateContents error. Error committing to file scenery-phet/scenery-phet-strings_da.json

> [{"stringKey":"FaceWithPointsNode.pattern_0sign_1points","stringValue":"{0}{1}"}]

**Comment by jbphet** [comment]
This issue was discussed in the 12/17/2015 developer meeting.  @phetsims/matt-p may potentially help work on it.

**Comment by jbphet** [comment]
I've verified that the email error message shown in https://github.com/phetsims/rosetta/issues/105#issuecomment-165829226 was indeed spurious by checking that the commit worked.  However, the timing looks odd.  This may help us figure out why this is happening (perhaps too many fast commits to GitHub in a row).  These commits were made just before the commit attempt that triggered the error report: 
+ https://github.com/phetsims/babel/commit/4b17b72143e994ba71412f4d1004ec3ff62e9b40
+ https://github.com/phetsims/babel/commit/b307f85111f28a56793c26a97ead803b98362311
+ https://github.com/phetsims/babel/commit/3c7af1510a9fdcd94399e1a56694b34f91280691.

...then comes the commit for which the error was reported:
+ https://github.com/phetsims/babel/commit/3eb275e0e9cfb71729e335ebd646a5a4cbc43d20

If you look at the timing of the commits, the last one happens *3 hours later* than the others, but rosetta's log shows them all happening at roughly the same time.  Could GitHub have a feature that somehow detects and slows down automatic commits?  At any rate, this is probably a clue for resolving this issue.

**Comment by jbphet** [comment]
Another incident where a "PUSH FAILED" email was sent but the commit actually worked occurred at 4:20 AM Jan 28 2016 for the file ```reactants-products-and-leftovers-strings_pt_BR.json```.  Here is the commit: https://github.com/phetsims/babel/commit/4a30e08e8d4749261c0e9269c9710144c2351f5d.

**Comment by jbphet** [comment]
There have been several more PUSH FAILED messages received over the last couple of weeks, probably around six total.  In some cases, it looks like the push really does fail, but the user continues to translate other simulations and eventually ends up pushing the (usually common) strings successfully.

**Comment by jbphet** [comment]
I just spent a little time looking at this, since we continue to get these PUSH FAILED messages on a regular basis.  When I look at the log, it appears that there are errors that occur on commits fairly frequently, but the 2nd attempt generally goes through, otherwise we'd be getting even more of these PUSH FAILED emails.  It seems like the cases where we actually have a full failure to commit and push, which means that two attempts failed, is when two *initial* attempts for different files fail one after another.

As I examine the code in commitQueue.js and the information in the build log, I can see that rosetta is attempting to do all of the commit in a batch, i.e. is doesn't what for one to finish before starting the next.  Maybe the problem is that GitHub isn't able to handle this many commits in rapid succession, which would make sense - it would be a hard thing to handle.  I think it would be worthwhile to modify our code to try serializing the individual commits, meaning that one needs to finish before the next one starts.

**Comment by jbphet** [comment]
We got quite a burst of PUSH FAILED messages between 1:26 AM and 1:58 AM on June 9 2016.  I looked through the log and didn't see anything unusual - just a bunch of failures when trying to push changes to GitHub.  My guess is that GitHub was doing some maintenance or having some issues at the time (we've seen it have problems for a while in the past).  @oliver-phet said that the translator contacted him about the issues.  We decided to tell the translator to try again and see if things work better.  If they do, I'll assume that it was a temporary glitch with GitHub.  If he runs into issues again, I'll look into whether something about the interaction with GitHub has changed.

**Comment by jbphet** [comment]
I received 17 of these PUSH FAILED notifications in the period from Wed Dec 21 2016 through 1:36 PM Wed Dec 28 2016.  This is far higher than the normal rate.  This may be due in part to the publication of the States of Matter and States of Matter Basics, which are popular sims and the full version was made available for translation on Dec 20 2016.

I have a hypothesis about this uptick in failures.  Six of the failures are for attempts to submit a change for the string "adaptedFrom" in `joist-strings_en.json`.  This string contains a carriage return character in most submissions, and I think it is incorrectly being identified by Rosetta as having changed every time a submit occurs, which is increasing the number of times that multiple files are submitted simultaneously, which is increasing the failure rate.  I'll log a separate issue about this.

**Comment by pixelzoom** [comment]
@jbphet said:
> Six of the failures are for attempts to submit a change for the string "adaptedFrom" in `joist-strings_en.json`.

Note that "adaptedFrom" was removed from joist-strings_en.json in https://github.com/phetsims/joist/issues/383.

**Comment by jbphet** [comment]
I spent some time digging into the cause of these PUSH FAILED messages, and I'm seeing what seems to me like a big architectural problem with Rosetta.  There is an async task queue called ```commitQueue``` that I thought - based on the name and the documentation - serialized commits going to GitHub so that they wouldn't all happen at once.  It doesn't.  It is actually a queue of translation submissions.  In other words, this queue queues up multiple submissions of translations so that they aren't processed at the same time.  This is a good thing, but it turns out that commits to GitHub are **not** being queued up and handled in sequence, and since each submission of a translation can cause changes to multiple string files (such as those for the main sim and those for common code), Rosetta often attempts multiple commits to GitHub in rapid succession, which is causing or at least contributing to failures of commit and push attempts.  I plan to add a second async queue to handle the GitHub commits, and in the process clean up the names, logging, and documentation of the current ```commitQueue``` so that it will be clearer what it does and how it works.

**Comment by samreid** [comment]
Trying to understand the preceding comment: would it solve the problem to make the `commitQueue` wait until a commit completes before attempting the next commit?

**Comment by jbphet** [comment]
Yes, absolutely.  That is essentially what an ```async.queue``` allows one to do.  There may be simpler ways, and I'll investigate as I work on this, but that's the solution that the original developer of the Rosetta code chose.

**Comment by jbphet** [comment]
I'm starting to implement a solution for this issue.  As a first step, I wanted to be able to make sure I can replicate the on the existing code.  Below are the steps that I went through to do this.  It was done on ox-dev, which is set up to push string changes to the "tests" branch of babel.  As a side note, we've been using the local "ab" (Abkhazian) because it's the first locale on the list, which makes testing easy, and because we've never received a translation in this language.  This is admittedly a little risky, and we may have a little cleanup to do if someone decides to start translating into this language.

Steps to Replicate:
--------------------
1. Go to the main translation page and log in.  For ox-dev, that's https://ox-dev.colorado.edu/translate/
2. Choose "Circuit Construction Kit: DC" as the sim to translate
3. Choose "Abkhazian (ab)" as the locale (should be the default)
4. Press the "Go" button
5. Under "Circuit Construction Kit: DC Strings", change the string "screen.lab" to a new value(it doesn't matter what)
6. Under "Common Strings", change the strings "menuItem.about" and "ammeter" to new values
7. Press the "Submit" button

This triggers creation of or changes to three string files: circuit-construction-kit-dc-strings_ab.json, joist-strings_ab.json, and circuit-construction-kit-common-strings_ab.json.  Trying to change these files at the same time triggers the problem where there are multiple requests going to GitHub in rapid succession and thus failing.  When I did this, here is what I saw:

![cck-ab](https://user-images.githubusercontent.com/3082831/33342596-e1583336-d43f-11e7-9ab3-3c9a2e88189b.PNG)

Looking at commits to the "test" branch in the babel repo, I can see that circuit-construction-kit-common-strings_ab.json was created (https://github.com/phetsims/babel/commit/8737ed6fd1ba1433222f2bdc223c85d62174fe25) and circuit-construction-kit-dc-strings_ab.json was created (https://github.com/phetsims/babel/commit/9eddf432c4e6de4a99838f95c1982d81ec673743), but the joist strings file was **not** created.  This behavior is essentially what was expected.

What the users generally do in this case is go back, make the changes for joist, and resubmit, and this works because then they are only changing one strings file.  I know this from looking at the logs.

I believe that if they subsequently went back and changed strings that were contained in 3 or more files, it would fail, but they wouldn't get the error message (but we would get the PUSH_FAILED email notification).

**Comment by jbphet** [comment]
I believe I have fixes in place for this, but I'm blocked from deploying it at the moment because I've also changed the build request (see https://github.com/phetsims/rosetta/issues/165), and that new build request isn't working on ox-dev and isn't deployed to phet-server.  I'm adding @mattpen as an assignee to remind him to let me know when I can test using ox-dev and when we can try deploying to phet-server.

**Comment by mattpen** [comment]
ox-dev is ready and available for testing.

@jbphet and I ran some tests on Friday, 1/5.

**Comment by jbphet** [comment]
This has been deployed to the production server for over a month now with no PUSH FAILED messages.  Closing.
